internal/controllers/s3bucket/handler.go

```diff
@@ -0,0 +1,80 @@
+package s3bucket
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	"github.com/opdev/subreconciler"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+
+	s3v1alpha1 "github.com/snapp-incubator/s3-operator/api/v1alpha1"
+	"github.com/snapp-incubator/s3-operator/internal/config"
+)
+
+// S3BucketReconciler reconciles a S3Bucket object
+type Reconciler struct {
+	client.Client
+	scheme  *runtime.Scheme
+	logger  logr.Logger
+	s3Agent *S3Agent
+	// reconcile specific variables
+	s3Bucket     *s3v1alpha1.S3Bucket
+	s3UserRef    string
+	s3BucketName string
+	rgwEndpoint  string
+}
+
+func NewReconciler(mgr manager.Manager, cfg *config.Config) *Reconciler {
+
+	return &Reconciler{
+		Client:      mgr.GetClient(),
+		scheme:      mgr.GetScheme(),
+		rgwEndpoint: cfg.Rgw.Endpoint,
+	}
+}
+
+//+kubebuilder:rbac:groups=s3.snappcloud.io,resources=s3buckets,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=s3.snappcloud.io,resources=s3buckets/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=s3.snappcloud.io,resources=s3buckets/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+
+	r.logger = log.FromContext(ctx)
+	r.s3Bucket = &s3v1alpha1.S3Bucket{}
+	r.s3BucketName = req.Name
+
+	// Get s3Bucket object
+	switch err := r.Get(ctx, req.NamespacedName, r.s3Bucket); {
+	case apierrors.IsNotFound(err):
+		r.logger.Info("CR not found.")
+		return ctrl.Result{}, nil
+	case err != nil:
+		r.logger.Error(err, "failed to fetch object")
+		return subreconciler.Evaluate(subreconciler.Requeue())
+	default:
+		r.s3UserRef = r.s3Bucket.Spec.S3UserRef
+		// Create a s3 session with the s3user credentials.
+		err = r.setS3Agent(ctx, req)
+		if err != nil {
+			r.logger.Error(err, "Failed to login on S3 with the user credentials")
+			return subreconciler.Evaluate(subreconciler.Requeue())
+		}
+		// Delete event
+		if r.s3Bucket.ObjectMeta.DeletionTimestamp != nil {
+			return r.Cleanup(ctx)
+		}
+	}
+
+	// Create event; Don't provision the object if it's in the ready status.
+	if r.s3Bucket.Status.Ready != true {
```

Comment:
We must not rely on status. The controller is expected to move the state toward the desired state (specified in spec) and set the status afterward.

Reply:
Every time the operator comes up (e.g., in case of a restart), it initiates a reconciliation, resulting in a provision and recreating existing buckets. I just put that condition to prevent the operator from that work, i.e., request to Ceph for the objects already in the ready status.

Reply:
Sending requests to check the status of the environment is an ordinary task for operators. Idempotency is preferred over efficiency.

Reply:
Thanks for the explanation, Ahmad jan.
I revised it.